File: 08. Trees/01. Binary Tree/05. Maximum depth in binary tree/main.js
class TreeNode {
    constructor(val) {
        this.val = val;
        this.left = this.right = null;
    }
}

// Recursive DFS approach
function maxDepthRecursive(root) {
    if (!root) return 0;
    return 1 + Math.max(maxDepthRecursive(root.left), maxDepthRecursive(root.right));
}

// Iterative BFS approach using Queue
function maxDepthIterative(root) {
    if (!root) return 0;

    let queue = [root];
    let depth = 0;

    while (queue.length > 0) {
        let size = queue.length;
        depth++;
        for (let i = 0; i < size; i++) {
            let node = queue.shift();
            if (node.left) queue.push(node.left);
            if (node.right) queue.push(node.right);
        }
    }

    return depth;
}

// Main function
function main() {
    let root = new TreeNode(3);
    root.left = new TreeNode(9);
    root.right = new TreeNode(20);
    root.right.left = new TreeNode(15);
    root.right.right = new TreeNode(7);

    console.log("Recursive Max Depth:", maxDepthRecursive(root));
    console.log("Iterative Max Depth:", maxDepthIterative(root));
}

// Run the main function
main();

File: 08. Trees/01. Binary Tree/05. Maximum depth in binary tree/main.py
class TreeNode:
    def __init__(self, val=0):
        self.val = val
        self.left = None
        self.right = None

# Recursive DFS approach
def maxDepthRecursive(root):
    if not root:
        return 0
    return 1 + max(maxDepthRecursive(root.left), maxDepthRecursive(root.right))

# Iterative BFS approach using Queue
from collections import deque
def maxDepthIterative(root):
    if not root:
        return 0
    
    queue = deque([root])
    depth = 0

    while queue:
        size = len(queue)
        depth += 1
        for _ in range(size):
            node = queue.popleft()
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)
    
    return depth

# Main function
def main():
    root = TreeNode(3)
    root.left = TreeNode(9)
    root.right = TreeNode(20)
    root.right.left = TreeNode(15)
    root.right.right = TreeNode(7)

    print("Recursive Max Depth:", maxDepthRecursive(root))
    print("Iterative Max Depth:", maxDepthIterative(root))

# Run the main function
main()
